static/src/javascripts/projects/common/utils/steady-page.js

```diff
@@ -0,0 +1,193 @@
+        return fastdom.read(function() {
+            // Add all the heights of the passed in batch
+            // removing the current height
+            return batch.filter(elementIsAbove).reduce(function(height, insertion) {
+                return height + readHeight(insertion.container);
+            }, 0);
+        });
+
+        function elementIsAbove(el) {
+            var parentElArray = [];
+            var parentEl = el.container;
+
+            // Push the container's parents into an array so we can calculate
+            // the position of the elements all the way until the body
+            while (parentEl) {
+                parentElArray.push(parentEl);
+                parentEl = parentEl.offsetParent;
+            }
+
+            var elTopPos = parentElArray.reduce(function(topPos, parentEl){
+                // This will loop up the parents until the body taking into account all the positions
+                // so that if an element is equal to the scrollY position then this will return 0
+                return topPos + parentEl.offsetTop + parentEl.clientTop;
+            }, 0);
+
+            // If the distance of the element from the top of the screen minus the height of the
+            // element we are measuring is less than the scroll position then we know that it will
+            // push the page down when loading in
+            return el.container.offsetHeight > -1 && (elTopPos - el.container.offsetHeight < window.scrollY);
+        }
+
+        function readHeight(el) {
+            var style = getComputedStyle(el);
+            var height = el.offsetHeight + parseInt(style.marginTop) + parseInt(style.marginBottom);
+            return isNaN(height) ? 0 : height;
+        }
+    }
+
+
+    /**
+     * Given a batch, call all of the callbacks on the insertion object
+     *
+     * @param  {Array} batch
+     */
+    function insertElements (batch) {
+        return fastdom.write(function () {
+            batch.forEach(function (insertion) {
+                insertion.cb();
+            });
+        });
+    }
+
+    /**
+     * Calculate the new height and either scroll or run the next batch of updates
+     *
+     * @param  {Number} state.height The total height of elements in the previous recursions
+     * @param  {Number} state.newHeight The height of all the elements added in the current recursion level
+     */
+    function calculateScrollY (state) {
+        if (q.empty()) {
+            // If the queue is empty (no more elements need to be added to the page) we immediately scroll
+            return state.newHeight + state.prevHeight + state.scrollY;
+        } else {
+            // If there are elements waiting to be added to the page we take the previous container's heights
+            // and recursively call the function so that we only scroll the page once the queue is empty -
+            // this prevents excessive and jarring scrolling
+            return go(assign(state, {
```

Comment:
I don't know if moving this function out of the promise chain made things any clearer, I think the recursive call to `go` again should be inside `go`